data_safe_haven/external/api/graph_api.py

```diff
@@ -573,12 +593,145 @@ def get_id_from_username(self, username: str) -> str | None:
                 next(
                     user
                     for user in self.read_users()
-                    if user["mailNickname"] == username
+                    if user["userPrincipalName"].split("@")[0] == username
                 )["id"]
             )
         except (DataSafeHavenMicrosoftGraphError, StopIteration):
             return None
 
+    def grant_role_permissions(
+        self,
+        application_name: str,
+        *,
+        application_role_assignments: Sequence[str],
+        delegated_role_assignments: Sequence[str],
+    ) -> None:
+        """Grant admin permission for requested application roles"""
```

Comment:
Granting roles to a service principal is done in two steps:

1. Assign the role to the application
2. Approve the request (this is granting permission)

Different RBAC permissions are needed for these two things. The second one will prompt for approval (which doesn't work on the command line) unless it has been approved by a Global Admin.

Reply:
OK, I think I see. Is "granting permissions" MS language? I think that is what confused me because "assigning a role" sounds very RBAC where as "granting permissions" sounds quite ad-hoc (give individual users, individual permissions).

I suppose I would be tempted to just call it `assign_role` in the function name and docstring, then use comments in the function to explain that in practice that is two steps (assignment and approval).

Does that sound right?

Reply:
Done